### PR TITLE
ConvertFrom-Json: Add -DateKind parameter

### LINKED ISSUE #13598: Add a -DateKind parameter to ConvertFrom-Json to control how System.DateTime / System.DateTimeOffset values are constructed
Follow-up from #13592:

## Summary of the new feature/enhancement

Introduce a new `-DateKind` parameter as an enumeration value as follows:

```
ConvertFrom-Json -DateKind { Local | Utc | Unspecified | Offset | None }
```

The goal is to parse timestamps in the input JSON as `[datetime]` /  `[datetimeoffset]` instances of the given kind - irrespective of the specific formatting of those timestamps.

* `Local` / `Utc` / `Unspecified` should create `[datetime]` instances with the given `.Kind` value.

  * A timestamp without time-zone information (UTC offset or `Z` suffix) should be assumed to be `Local` (a pitfall is that on calling `.ToLocalTime()` on something that was parsed as `Unspecified`, it is assumed to be a UTC value - `[datetime]::new(<unspecified-datetime>.Ticks, 'Local')` is the solution).
  * If `Unspecified` is requested, any UTC offset / `Z` suffix should be ignored.
  
* `Offset` should create a `[datetimeoffset]` instance instead, which preserves the exact UTC offset as specified or implied in the input.

  * A timestamp without time-zone information (UTC offset or `Z` suffix) should be assumed to be a caller-local timestamp.

* `None` should not create a date a `[datetime]` /  `[datetimeoffset]` at all and instead pass the _original string values_ through.

Note that on _serializing_, with `ConvertTo-Json`, both `Utc` and `Local` `[datetime]` instances serialize with the appropriate UTC offset (e.g., `"2020-09-08T16:24:21.577822-04:00"`), whereas `Unspecified` serializes without an offset (e.g., "2020-09-08T16:24:21.577822").

**Motivation**:

Since v6, the Json.NET-based `ConvertTo-Json` automatically deserializes _strings_ that contain an `"o"`-formatted (roundtrip format) date/time string (e.g., `"2020-09-07T09:44:13.769Z"`) or a _prefix_ of it that at least includes everything up to the _seconds_ part (e.g., `"2020-09-07T09:44:13:`) as `[datetime]` instances.

However, the resulting `[datetime]` instances' `.Kind` property (`Local`, `Utc`, or `Unspecified`) is determined by the specifics of the string value as follows:

* `Unspecified`, if there is no time-zone information in the input string.
* `Utc`, if the time-zone information is a trailing `Z`
* `Local`, if the time-zone information is given as a trailing UTC _offset_ (e.g., `+02:00`), in which case the offset is properly converted to the caller-local equivalent.

Note that the new `System.Text.Json` API that is being considered as the future underpinning of the `*-Json` cmdlets - see #11198 - exhibits the same behavior.

There are two problems with this behavior:

* A `Utc` instance _gives no indication that it is a UTC timestamp_ in its default output formatting (applies both to PowerShell's formatting and `.ToString()` - verify with  `[datetime]::UtcNow` and `[datetime]::UtcNow.ToString()`).

* Specific UTC offsets in the input are lost, because they are translated to a _caller-local_ `Local` instance.

### PR DESCRIPTION
**Pull request by jborean93**
# PR Summary

Adds the -DateKind parameter to the ConvertFrom-Json that allows the caller to control how DateTime strings are converted into an object. The default behaviour is to create a DateTime value with the Kind being Unspecified if no TZ is set, Utc if the TZ Z is set, Local (after conversion) if an explicit TZ is set. This adds a Utc, Local to explicitly set the Kind as desired as well as a Offset and String value to create a DateTimeOffset or keep as a string.

## PR Context

Provides the user the ability to create a `DateTimeOffset` value so that the underlying time zone information is preserved. Also gives the user the ability to keep it as a string or a specific `DateTime` with the `Local` or `Utc` kind.

Fixes: https://github.com/PowerShell/PowerShell/issues/13598

Docs PR for this change: https://github.com/MicrosoftDocs/PowerShell-Docs/pull/10737

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/pull/10737
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by iSazonov** [comment]
@mklement0 Could please review and confirm this work as you expect?

**Review by iSazonov — APPROVED** [review]
LGTM.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs [hunk 1]
```diff
@@ -151,16 +151,68 @@ public static object ConvertFromJson(string input, bool returnHashtable, out Err
         /// if the <paramref name="returnHashtable"/> parameter is true.</returns>
         [SuppressMessage("Microsoft.Naming", "CA1704:IdentifiersShouldBeSpelledCorrectly", Justification = "Preferring Json over JSON")]
         public static object ConvertFromJson(string input, bool returnHashtable, int? maxDepth, out ErrorRecord error)
+            => ConvertFromJson(input, returnHashtable, maxDepth, jsonDateKind: JsonDateKind.Local, out error);
```
Comment (mklement0) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:154]: Shouldn't this be `JsonDateKind.Default`?

**Comment by mklement0** [comment]
Thanks, @iSazonov  - I have one small question (see above), but otherwise it looks great. Thank you for tackling this, @jborean93.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs (hunk 1 above)
Comment (jborean93) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:154]: Yes it should be, sorry was playing around with comparing the defaults and never fixed this up when finalizing on the values.

Comment (jborean93) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:154]:
```suggestion
            => ConvertFromJson(input, returnHashtable, maxDepth, jsonDateKind: JsonDateKind.Default, out error);
```

**Review by SteveL-MSFT — APPROVED** [review]
LGTM, great addition

**Comment by iSazonov** [comment]
@jborean93 Thanks for your contribution!

**Comment by jborean93** [comment]
Thanks for the review everyone!

**Pull request merged**